src/dmd/todt.d

```diff
@@ -478,6 +478,92 @@ extern (C++) void Expression_toDt(Expression e, ref DtBuilder dtb)
         mem.xfree(q);
     }
 
+    static void evalHash(scope Expression key, scope Type keyType, scope out Expression hash_result)
+    {
+        scope Expression call;
+        // need to get the key-type hash function.
+        FuncDeclaration fd_tohash = null;
+
+        // if it's a struct search for the toHash()
+        // this is an optimisation as it means cheap expressionSemantic
+        if (auto st = keyType.isTypeStruct())
+        {
+            import dmd.id;
+            auto keyTypeStructDecl = st.sym;
+            assert(keyTypeStructDecl);
+            auto toHash = keyTypeStructDecl.search(key.loc, Id.tohash);
+            fd_tohash  = toHash ? toHash.isFuncDeclaration() : null;
+        }
+
+        if (fd_tohash)
+        {
+            scope dotvar = new DotVarExp(key.loc, key, fd_tohash);
+            call = new CallExp(key.loc, dotvar);
+        }
+        else
+        {
+            // didn't find a to-hash let's use hashOf from druntime.internal.hash
+            auto loadCoreInternalHash()
+            {
+                // TODO factor this with expressionsem.d into load runtime module
+                import dmd.dmodule;
+                import dmd.dimport;
+                import dmd.id;
+                import dmd.identifier;
+                import dmd.dsymbolsem;
+                __gshared Import impCoreInternalHash = null;
+                __gshared Identifier[2] coreInternalID;
+                if (!impCoreInternalHash)
+                {
+                    coreInternalID[0] = Id.core;
+                    coreInternalID[1] = Identifier.idPool("internal");
+                    auto s = new Import(key.loc, coreInternalID[], Identifier.idPool("hash"), null, false);
+                    // Module.load will call fatal() if there's no std.math available.
+                    // Gag the error here, pushing the error handling to the caller.
+                    uint errors = global.startGagging();
+                    s.load(null);
+                    if (s.mod)
+                    {
+                        s.mod.importAll(null);
+                        s.mod.dsymbolSemantic(null);
+                    }
+                    global.endGagging(errors);
+                    impCoreInternalHash = s;
+                }
+                return impCoreInternalHash.mod;
+            }
+
+            auto se = new ScopeExp(key.loc, loadCoreInternalHash());
+            import dmd.identifier;
+            scope dotid = new DotIdExp(key.loc, se, Identifier.idPool("hashOf"));
+            call = new CallExp(key.loc, dotid, key);
+        }
+
+        import dmd.expressionsem;
+        import dmd.dinterpret;
+        import dmd.ctfeexpr : exceptionOrCantInterpret;
+        if (call)
```

Comment:
can be an `assert`, since `call` is assigned in both branches.

Reply:
true